src/c++/vdo/base/vdo.c

```diff
@@ -500,7 +940,48 @@ static int initialize_vdo(struct vdo *vdo, struct device_config *config,
 	vdo_initialize_completion(&vdo->admin.completion, vdo, VDO_ADMIN_COMPLETION);
 	init_completion(&vdo->admin.callback_sync);
 	mutex_init(&vdo->stats_mutex);
-	result = read_geometry_block(vdo);
+
+	// Check to see whether we should format the device
+	result = vdo_is_formatted(vdo, &formatted);
+	if (result != VDO_SUCCESS) {
+		*reason = "Could not determine whether layout is empty";
+		return result;
+	}
+
+	// This is here to distiguish between regular error and not formattable.
+	if (formatted == 0) {
+		/* Get size of future indexer */
+		block_count_t index_blocks;
+
+		result = compute_index_blocks(config, &index_blocks);
+		if (result != VDO_SUCCESS) {
+			*reason = "Unable to calculate index size";
+			return VDO_BAD_CONFIGURATION;
+		}
+		config->index_blocks = index_blocks;
+
+		/* Check if minimal size of the VDO is too big */
+		block_count_t slab_blocks = 1 << config->slab_bits;
+		block_count_t fixed_layout_size = 2 + index_blocks +
+			DEFAULT_VDO_BLOCK_MAP_TREE_ROOT_COUNT +
+			DEFAULT_VDO_RECOVERY_JOURNAL_SIZE + VDO_SLAB_SUMMARY_BLOCKS;
+		block_count_t necessary_size = fixed_layout_size + slab_blocks;
+
+		if (necessary_size > config->physical_blocks) {
+			u64 minimumSize = necessary_size * VDO_BLOCK_SIZE;
+
+			vdo_log_error("Minimum required size for VDO volume: %llu bytes",
+				      (unsigned long long) minimumSize);
+			*reason = "Not enough space to make a VDO";
+			return VDO_NO_SPACE;
+		}
+
+		result = vdo_format(vdo, reason);
+		if (result != VDO_SUCCESS)
+			return result;
+	}
+
+	result = load_geometry_block(vdo);
```

Comment:
We already read the geometry block. Or if we've formatted, the geometry info is in memory already. There's no reason to read the block again.

Reply:
I did it for safety sake. In this way. we read what is written to disk, not what was in memory.